Auth 2.0 Iteration 2

HaoK opened the pull request:
Test showing dynamic add/remove of schemes: https://github.com/aspnet/Security/pull/1113/commits/d92d92384f7f6cf2e5e17c975efaa0bd6ab299c0#diff-220317c67a0ea163cf751157519c7f66R29

- Will move this to HttpAbstractions once its baked (maybe the base Authentication package will need to move as well for servers to use the common infrastructure).

cc @Eilon @Tratcher @davidfowl

Follow up items:
- Revisit HandleRequest routing
- Normalize/revisit all of the events (some derive from Control some don't), mixed handling patterns.
- Scheme options validation (things like SignInScheme defaults are initalized in handler, so can't validate on ConfigureTime)

Test holes:
- Need unit tests for base OAuth and new Auth service interfaces, and when UseAuth is missing (RemoteAuth should throw if feature is missing)

dnfclas commented:
Hi __@HaoK__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by .NET Foundation and real humans are currently evaluating your PR.</p>

TTYL, DNFBOT;

HaoK commented:
Rather than nuking the existing auth package requiring updating everything at once, implementation of the new stack will be in a Authentication.Impl package until everything is done

- Renamed IAuthenticationSchemeHandler => IAuthenticationHandler (taking over the old interface name), shouldn't lose any clarity
- Added IAuthenticationHandlerResolver which has a single method which resolve handler instance for a given scheme + request.

brockallen commented:
@leastprivilege and I have a TODO item to look over and give feedback. We were hoping to have done this last week, but it didn't happen. Perhaps this weekend or next week. Thanks.

HaoK commented:
Sure, you guys might want to look at the old PR first for better context/info, and it has more or less the same shape, it will take a bit for me to stitch everything back together again in this PR.

HaoK commented:
Seems too confusing to have any dependency on the old HttpAuthentication feature, going to copy ChallengeBehavior and AuthenticationProperties instead, since we aren't supporting old and new together.

HaoK commented on test/Microsoft.AspNetCore.Authentication.Test/Cookies/CookieMiddlewareTests.cs:
```diff
@@ -851,68 +740,85 @@ private Task SignOutAsWrong(HttpContext context)
 
         [Fact]
-        public async Task MapWillNotAffectChallenge()
+        public async Task MapWillAffectChallenge()
```
Note: this is a behavior change from old stack to new stack (Map does affect redirect url for challenge now that its not tied to any middleware)

PinpointTownes replied:
Be careful, it's most likely not a wanted change: if `/page` is configured as the callback path and the user is actually redirected to `/login/page`, all he'll get is a 404 response, as no login page will handle the request.

I assume it similarly impacts all the RAT-based middleware that use callback paths and have built-in endpoints? (OIDC, OAuth2, Twitter...)

HaoK commented:
Cookies and tests have been updated to use the new stack, and pull system clock from DI

PinpointTownes commented on test/Microsoft.AspNetCore.Authentication.Test/Cookies/CookieMiddlewareTests.cs:
```diff
@@ -851,68 +740,85 @@ private Task SignOutAsWrong(HttpContext context)
         }
 
-        [Fact]
+        [ConditionalFact(Skip = "Revisit, exception no longer thrown")]
```
Hum, why? The `InvalidOperationException` is there to inform the dev' that something is wrong with his setup. I don't think it's preferable to silently ignore the operation.

HaoK replied:
I think the old test threw because it was doing an AutomaticChallenge followed by an second challenge, which no longer happens in the new stack, I'll have to debug the old test, to see what was triggering it

HaoK commented on test/Microsoft.AspNetCore.Authentication.Test/Cookies/CookieMiddlewareTests.cs:
```diff
@@ -739,11 +739,19 @@ private Task SignOutAsWrong(HttpContext context)
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
+        public async Task MapWillAffectChallengeOnlyWithUseAuth(bool useAuth)
```
FYI @PinpointTownes fixed the map + resolve url behavior to work similar to before (using the path of the single middleware), via a new IAuthenticationFeature like we discussed @Tratcher

PinpointTownes replied:
Really great news!

Tratcher commented on samples/OpenIdConnectSample/Properties/launchSettings.json:
```diff
@@ -3,8 +3,8 @@
     "windowsAuthentication": false,
     "anonymousAuthentication": true,
     "iisExpress": {
-      "applicationUrl": "http://localhost:42023",
```
revert, these samples need to stay on a specific port.

Tratcher commented on src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationScheme.cs:
```diff
@@ -0,0 +1,34 @@
+    {
+        public AuthenticationScheme(string name, Type handlerType, Dictionary<string, object> settings)
+        {
+            Name = name ?? throw new ArgumentNullException(nameof(name));
```
FYI: @Eilon has rejected the use of this syntax.

blowdart replied:
Does that catch string.empty?

Tratcher commented on src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationScheme.cs:
```diff
@@ -0,0 +1,34 @@
+{
+    public class AuthenticationScheme
+    {
+        public AuthenticationScheme(string name, Type handlerType, Dictionary<string, object> settings)
```
Why not just take an IReadOnlyDictionary?

HaoK replied:
Good point, fixed

Tratcher commented on src/Microsoft.AspNetCore.Authentication.Abstractions/ChallengeBehavior.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace Microsoft.AspNetCore.Authentication
+{
+    public enum ChallengeBehavior
```
Looking to eliminate this. Let Challenge and Forbidden be separate code paths, and Automatic would be resolved at a higher level.

HaoK replied:
Yeah but we likely can't do this until there's a bunch of corresponding AuthZ changes, we should tackle that in a follow on PR specifically for AuthZ (see https://github.com/aspnet/Security/issues/901)

Tratcher commented on src/Microsoft.AspNetCore.Authentication.Abstractions/Microsoft.AspNetCore.Authentication.Abstractions.csproj:
```diff
@@ -0,0 +1,22 @@
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="Microsoft.AspNetCore.DataProtection" Version="1.2.0-*" />
+    <PackageReference Include="Microsoft.AspNetCore.Http" Version="1.2.0-*" />
```
What did you need from .HTTP? It's unusual for an Abstractions package to have this many concrete dependencies.

HaoK replied:
Good catch, these were just carried over from eixsting Authentication.csproj, I switched them to just the Http.Abstractions package and things were fine.

Tratcher commented on src/Microsoft.AspNetCore.Authentication.Cookies/CookieAuthenticationOptions.cs:
```diff
@@ -2,19 +2,15 @@
-using Microsoft.Extensions.Options;
 
-namespace Microsoft.AspNetCore.Builder
+namespace Microsoft.AspNetCore.Authentication.Cookies
```
We'd used this namespace to improve intellisence in Startup, it's the same namespace the extensions were in. Should these move to the DI namespace?

HaoK replied:
I don't think that's necessary with lambdas, as intellisense knows the type from the signature, the namespace hack was only needed for `new`

Tratcher commented on src/Microsoft.AspNetCore.Authentication.Cookies/Events/BaseCookieContext.cs:
```diff
@@ -23,5 +24,7 @@ public class BaseCookieContext : BaseContext
 
         public CookieAuthenticationOptions Options { get; }
+
+        public AuthenticationScheme Scheme { get; }
```
Would Scheme go on BaseAuthenticationContext?

HaoK replied:
I tried to bridge the context's between the service and the handlers.  The service doesn't care about the scheme object, only the string, but maybe we should just have a BaseAuthenticationHandlerContext that does have the scheme.  I have a todo in the top comment to make a pass through all of our events/context objects since they have always been a bit of a hodgepodge mess

Tratcher commented on src/Microsoft.AspNetCore.Authentication.Facebook/FacebookMiddleware.cs:
```diff
@@ -1,89 +0,0 @@
-                throw new ArgumentNullException(nameof(options));
-            }
-
-            if (string.IsNullOrEmpty(Options.AppId))
```
These checks are missing

HaoK replied:
Grrr yup, I'll add a test for these as well

HaoK replied:
Moved the checks to Validate in the options which at least causes an exception, but the current behavior isn't perfect, in that it blows up the first time the scheme is used in the auth service.

Tratcher replied:
Weren't we going to call Validate from the UseFooAuth extension too?

HaoK replied:
Sorry I mispoke, it blows up when the Auth service is constructed as the schemes are built when SchemeProvider/HandlerResolvers are built.  So maybe that's not that bad if the auth service itself cannot be resolved due to a configuration error.

HaoK replied:
But the callstack is huge and messy but at least the error message is clear in the argument exception.

Tratcher replied:
Can you share the callstack/exception?

HaoK replied:
```
typeof(System.ArgumentException): The 'ClientId' option must be provided.
Parameter name: ClientId
Stack Trace:
   at Microsoft.AspNetCore.Authentication.OAuth.OAuthOptions.Validate() in C:\GitHub\Security\src\Microsoft.AspNetCore.Authentication.OAuth\OAuthOptions.cs:line 33
   at Microsoft.Extensions.DependencyInjection.AuthenticationServiceCollectionExtensions.<>c__DisplayClass2_0`2.<AddSchemeHandler>b__1(AuthenticationSchemeBuilder b) in C:\GitHub\Security\src\Microsoft.AspNetCore.Authentication.Impl\AuthenticationServiceCollectionExtensions.cs:line 66
   at Microsoft.AspNetCore.Authentication.AuthenticationOptions.AddScheme(String name, Action`1 configureBuilder) in C:\GitHub\Security\src\Microsoft.AspNetCore.Authentication.Abstractions\AuthenticationOptions.cs:line 42
   at Microsoft.Extensions.DependencyInjection.AuthenticationServiceCollectionExtensions.<>c__DisplayClass2_0`2.<AddSchemeHandler>b__0(AuthenticationOptions o) in C:\GitHub\Security\src\Microsoft.AspNetCore.Authentication.Impl\AuthenticationServiceCollectionExtensions.cs:line 56
   at Microsoft.Extensions.Options.OptionsCache`1.CreateOptions()
   at System.Threading.LazyInitializer.EnsureInitializedCore[T](T& target, Boolean& initialized, Object& syncLock, Func`1 valueFactory)
   at Microsoft.Extensions.Options.OptionsCache`1.get_Value()
   at Microsoft.AspNetCore.Authentication.DefaultAuthenticationSchemeProvider..ctor(IOptions`1 options) in C:\GitHub\Security\src\Microsoft.AspNetCore.Authentication.Impl\DefaultAuthenticationSchemeProvider.cs:line 18
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.<>c__DisplayClass17_0.<RealizeService>b__0(ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
   at Microsoft.AspNetCore.Authentication.AuthenticationHttpContextExtensions.ChallengeAsync(HttpContext context, String scheme, AuthenticationProperties properties, ChallengeBehavior behavior) in C:\GitHub\Security\src\Microsoft.AspNetCore.Authentication.Abstractions\AuthenticationHttpContextExtensions.cs:line 23
   at Microsoft.AspNetCore.Authentication.AuthenticationHttpContextExtensions.ChallengeAsync(HttpContext context, String scheme, AuthenticationProperties properties) in C:\GitHub\Security\src\Microsoft.AspNetCore.Authentication.Abstractions\AuthenticationHttpContextExtensions.cs:line 20
   at Microsoft.AspNetCore.Authentication.AuthenticationHttpContextExtensions.ChallengeAsync(HttpContext context, String scheme) in C:\GitHub\Security\src\Microsoft.AspNetCore.Authentication.Abstractions\AuthenticationHttpContextExtensions.cs:line 17
   at Microsoft.AspNetCore.Authentication.OAuth.OAuthTests.<>c__DisplayClass0_0.<ThrowsIfClientIdMissing>b__4() in C:\GitHub\Security\test\Microsoft.AspNetCore.Authentication.Test\OAuthTests.cs:line 33
```

Tratcher replied:
That only says OAuth, it doesn't tell you Facebook. And Facebook calls it AppId. This is going to cause confusion.

HaoK replied:
I copied a different exception failure for the above example from the oauth tests, the facebook one says app id and has a facebookoptions.validate at the top.

Tratcher replied:
Ah, good to know

HaoK commented on src/Microsoft.AspNetCore.Authentication.Cookies/CookieAuthenticationHandler.cs:
```diff
@@ -216,20 +216,10 @@ private CookieOptions BuildCookieOptions()
         {
             // Only renew if requested, and neither sign in or sign out was called
-            if (!_shouldRefresh || SignInAccepted || SignOutAccepted)
+            if (!_shouldRefresh || SignInCalled || SignOutCalled)
```
Test hole here, need to add a test that verifies cookies aren't renewed if SignIn or SignOut is called. (this was broken before this change)

HaoK commented:
We also had lost the OauthOptions validations too which had no test coverage :/  I added some basic validation tests for those, also compact'ed the test structure, I nuked all the folders that had a single file, so basically only OpenIdConnect has a subfolder in Auth.Test now.

brockallen commented:
Hi all -- again, no time yet to have a look, but I wanted to let you know how we're integrating with the current security middleware in IdentityServer. In short, we're registering our own authentication middleware to be able to do a decorator pattern like feature to applications hosting us. We'd like to know if we can do something similar in this v2 work. Here's the main handler we have now:

https://github.com/IdentityServer/IdentityServer4/blob/dev/src/IdentityServer4/Hosting/AuthenticationHandler.cs

PinpointTownes commented:
@brockallen the "authentication middleware" chain is no longer a thing in this prototype, so you won't be able to port your handler as-is. That said, I guess you could achieve something similar using the new `IAuthenticationService`, which is more or less the equivalent of `AuthenticationManager`, but DI-injected.

brockallen commented:
> I guess you could achieve something similar by replacing IAuthenticationService

Right, but then we'd probably need to decorate whatever implementation is provided by default. Too bad the core DI system is so lacking in so many useful features.

PinpointTownes commented:
> Too bad the core DI system is so lacking in so many useful features.

There a few workarounds to make the default DI stack (kinda) work with the decorator pattern, but it's indeed not ideal... I guess that's the price to pay for having full DI here :trollface: 

> To be honest, I'm not a huge fan of the "full DI" approach suggested by @HaoK and @brockallen as it will likely introduce a whole new set of issues without really solving the initial one.

https://github.com/aspnet/Security/issues/1061#issuecomment-266092809

brockallen commented:
Well, the original chained handler design in katana was essentially an attempt at designing a poor man's decorator pattern in a system that had no DI. Moving to a service oriented approach in a DI system with missing features is just changing the surrounding code. I think I agree that it's not addressing the fundamental requirement. I've yet to look into the other changes to see if they've done anything to make the overall design better.

HaoK commented:
@brockallen it looks like you could just replace the DefaultAuthenticationService with your own derived one, something like this:

```
  public class AuthenticationService : DefaultAuthenticationService, IAuthenticationService 
  {
        private readonly IdentityServerOptions _options;
        private readonly IHttpContextAccessor _context;
        private readonly ISessionIdService _sessionId;
        private readonly ILogger<AuthenticationHandler> _logger;
        private readonly IEventService _events;

        public AuthenticationService(
            IAuthenticationHandlerResolver cache, 
            IClaimsTransformation transform,
            IHttpContextAccessor context, 
            IdentityServerOptions options, 
            ISessionIdService sessionId,
            IEventService events,
            ILogger<AuthenticationService> logger) : base(cache, transform)
        {
            _context = context;
            _options = options;
            _sessionId = sessionId;
            _events = events;
            _logger = logger;
        }

        public override async Task SignInAsync(SignInContext context)
        {
            if (context.AuthenticationScheme == _options.Authentication.EffectiveAuthenticationScheme)
            {
                AugmentContext(context);
                await RaiseSignInEventAsync(context.Principal);
            }
            await base.SignInAsync(context);
        }

        private void AugmentContext(SignInContext context)
        {
            _logger.LogDebug("Augmenting SignInContext");

            context.Principal.AssertRequiredClaims();
            context.Principal.AugmentMissingClaims();

            _sessionId.CreateSessionId(context);
        }

        public override async Task SignOutAsync(SignOutContext context)
        {
            if (context.AuthenticationScheme == _options.Authentication.EffectiveAuthenticationScheme)
            {
                _sessionId.RemoveCookie();
                await RaiseSignOutEventAsync();
            }
            await base.SignOutAsync(context);
        }

        // Would need to move to your own middleware most likely
        internal async Task InitAsync()
        {
            var auth = GetAuthentication();
            _handler = auth.Handler;
            auth.Handler = this;

            await _sessionId.EnsureSessionCookieAsync();
        }

        internal void Cleanup()
        {
            var auth = GetAuthentication();
            auth.Handler = _handler;
        }

        // No longer needed, DI registration
        IHttpAuthenticationFeature GetAuthentication()
        {
            var auth = _context.HttpContext.Features.Get<IHttpAuthenticationFeature>();
            if (auth == null)
            {
                auth = new HttpAuthenticationFeature();
                _context.HttpContext.Features.Set(auth);
            }
            return auth;
        }

        private async Task RaiseSignInEventAsync(ClaimsPrincipal principal)
        {
            await _events.RaiseLoginEventAsync(principal);
        }

        private async Task RaiseSignOutEventAsync()
        {
            var principal = await _context.HttpContext.GetIdentityServerUserAsync();
            if (principal.IsAuthenticated())
            {
                await _events.RaiseLogoutEventAsync(principal);
            }
        }
    }
   }
```

brockallen commented:
Sure, but given our design and the design of the new auth service, who knows if the hosting app has chosen to do their own... so we'd need to decorate whatever is already registered. But in any event, good news that it won't be all that different that what we're faced with today. Thanks.

HaoK commented:
Yeah so the idea is the old `IAuthenticationHandler` is left alone for now and will be unchanged, the new stack basically will ignore the old stack, and can live alongside the old, they don't interact with each other, so you could potentially support both the old and new at the same time if you desired.

HaoK commented:
Updated adding a test exercising dyanmically adding and removing schemes, was able to remove the ugly ResolverFunc with new logic that just tries to resolve the service against RequestServices, and if that fails uses ActivatorUtilities.CreateInstance instead.  Which means the handlers don't have to be registered in advance unless they want to have a non scoped lifetime.

HaoK commented:
Added new flag on AuthenticationScheme.CanHandleRequests which is what the AuthenticationMiddleware will look at to build the list of handlers to give a chance during the request cycle.  RemoteAuth handlers are set to true (JwtBearer/Cookies set this flag to false).

Tratcher commented:
Most of the middleware handle requests on a specific route. Should middleware register a route with the service?

PinpointTownes commented on src/Microsoft.AspNetCore.Authentication.Impl/AuthenticationServiceCollectionExtensions.cs:
```diff
@@ -46,8 +45,7 @@ public static IServiceCollection AddAuthentication(this IServiceCollection servi
 
-        // REVIEW: rename to just AddScheme?
-        public static IServiceCollection AddSchemeHandler<TOptions, THandler>(this IServiceCollection services, string authenticationScheme, Action<TOptions> configureOptions)
+        public static IServiceCollection AddScheme<TOptions, THandler>(this IServiceCollection services, string authenticationScheme, Action<TOptions> configureOptions, bool canHandleRequests)
```
What's the point of doing that? Request handling is something the handlers do by implementing `HandleRequestAsync`, it shouldn't be configurable. Specially by the end user.

HaoK replied:
This is the infrastructure method.  AddGoogle/Cookies/FacebookAuth use this method to register themselves and set the flag.  The idea was to make it opt in for the schemes to get called by the middleware, but as @Tratcher mentioned, next I'll probably scope the blanket callback the handlers get today, and instead they would register their callback paths and the middleware would do some ghetto routing, instead of this opt in bit..  You'd still be able to do more advanced routing by replacing the AuthenticationMiddleware (or adding another customer middleware).  The motivation is to avoid creating the handlers for all the schemes until they are needed/called.

PinpointTownes replied:
This could potentially lead to really weird issues if the users directly called `AddScheme()` without using the appropriate `canHandleRequests`, but since it's justified by perf', why not... :sweat_smile:

PinpointTownes replied:
Please don't make your custom routing logic too hard to override.

We use a specific `MatchEndpoint` event in ASOS that allows users to dynamically determine the type of request/endpoint, and I really want to preserve it in the new world: https://github.com/aspnet-contrib/AspNet.Security.OpenIdConnect.Server/blob/dev/src/AspNet.Security.OpenIdConnect.Server/OpenIdConnectServerHandler.cs#L27-L170.

HaoK replied:
The new logic will be pushed shortly, I'm just making sure all the tests still work...  

The default routing is basically RemoteOptions single CallbackPath has become many CallbackPaths(to support OIDC's two). canHandleRequests is gone, but there's a new helper extension called AddRemoteScheme, that adds the callback paths to the scheme.

The middleware will call the IAuthenticationSchemeProvider's GetRequestHandlerSchemes with the request path to get back the list of schemes to call ProcessRequest on.  The default implementation just maps the path string to schemes.  

Looking the ASOS, you would just need to add them all as callback events, and should be good to go I think.

PinpointTownes replied:
> Looking the ASOS, you would just need to add them all as callback events, and should be good to go I think.

You mean as callback path? In this case, this won't work. The `MatchEndpoint` event MUST be called for every request to allow the developer to decide whether it's a request that should be handled or not, depending on his own policy (not just the path).

Not to mention that I ultimately want to support URLs like /connect/register/[dynamic routing parameter]).

HaoK replied:
So right now you can accomplish that with a dedicated custom ASOS middleware that always calls into that schemes process request, or you can override the IAuthenticationSchemeProvider's GetRequestHandlerScheme method to always return the ASOS scheme for all requests.

PinpointTownes replied:
Replacing `IAuthenticationSchemeProvider` is definitely not an option (it will be broken if the user decides to register his own instance).

Having to create a custom middleware just for that seems weird. I thought the whole point of this PR was to remove the "authentication middleware" concept (except the new global one, of course)? :open_mouth:

brockallen replied:
> Replacing IAuthenticationSchemeProvider is definitely not an option (it will be broken if the user decides to register his own instance).

So many workarounds in various ASP.NET features due to a lack of a robust DI system.

HaoK replied:
The idea was to move Auth to be a service, which has more or less been accomplished.  We wanted to move away from invoking all handlers for all requests by default (for perf and to work better with multi-tenancy), while not blocking the ability for a scheme to inspect all requests.

If replacing the default `IAuthenticationSchemeProvider` is not an option, then you are only left with adding your own call to `IAuthenticationSchemeProvider.GetSchemeAsync("ASOS").ProcessRequestAsync()` somewhere, probably via middleware

If this is ends up being a common scenario, we can always add a new options flag like InvokeAllSchemesForAllRequests (aka SlowMoMode = true) for the single Auth middleware to call them all like before.

HaoK replied:
Or maybe we just need a slightly less ghetto routing point that's separate from IAuthenticationSchemeProvider so you can replace that without affecting other things.

PinpointTownes replied:
Sounds like a real PITA just for an hypothetical perf' gain that wasn't even measured 😄

HaoK replied:
Well its kind of required for this https://github.com/aspnet/Security/issues/1132 once we commit to letting people add hundreds/thousands of schemes, we can't really call them all for every request.

PinpointTownes replied:
Why would you add hundreds/thousands of schemes? It doesn't make much sense, even for multi-tenant scenarios (in this case, you'd want a single scheme handler, but with potentially unlimited options, e.g coming from a DB)

brockallen replied:
> Why would you add hundreds/thousands of schemes

I have a customer that uses ws-fed with ~2K external STSs. We could not do it with Katana, so the protocol work had to be done by hand.

PinpointTownes replied:
By scheme, I mean scheme handler. In your case, you'd have only one scheme handler (WS-Fed), with 2K different sets of options, resolved dynamically.

HaoK replied:
The scheme settings are all singletons right now, so if each tenant has a different ClientId/Secret, they will need their own scheme.

HaoK replied:
We have a whole new can of worms once we try to enable reading options from the DB, as we have no async, and lots of our assumptions (options being singleton) go out the window as soon as we start to do dynamic per request options binding (against databases no less...and we are using super slow reflection binding code against configuration as well)...

PinpointTownes replied:
@HaoK point taken, tho' I'm not sure why you'd want to support scenarios with "hundreds/thousands of schemes" if there's no real support in the options stack (I'm not sure having hundreds of schemes + hundreds of handlers + hundreds of options instances is really _scalable_).

Not being able to invoke my handler for every path is really annoying for me. As a compromise, would you consider bringing back your per-scheme "promiscuous mode"? (aka `CanHandleRequest` :smile:)

HaoK commented:
How does something like `IEnumerable<PathString> CallbackPaths` on AuthenticationScheme (most would only have one, OIDC has two for sign in + sign out).  They don't have to be unique, and the middleware will do the routing to the all of the appropriate scheme's HandleRequest that matched.  We'd still keep the bool signifying whether we should stop processing the request.

HaoK commented:
Updated: schemes must now explicitly register callback paths in their scheme data.  Schemes ProcessRequest will only be called if Request.Path matches a callback path they specify.

Will probably remove HandleRemoteAuthenticate since its no longer needed.  Also will change the base ProcessRequest to throw now that it shouldn't be invoked anymore

Tratcher commented:
How terrible would it be to pull in actual routing?

HaoK commented:
Well we could add something like "routes.MapAuthenticationRoutes()".  We'd still have to have some internal plumbing in the schemes to store the routes for this method to consume.  So the required startup pattern would then become:

```
    app.UseRouter(routes => routes.MapAuthenticationRoutes());
    app.UseAuthentication();
```

Doesn't look too unreasonable, but is it reasonable for us to require routing for auth to work?

Tratcher commented:
Not bad. How well would that work with dynamically added/removed auth schemes?

HaoK commented:
I'm not sure, depends on how well routing supports dynamically adding routes :)  Let me see

PinpointTownes commented:
Forking my comment as it was hidden by GitHub:

@HaoK point taken, tho' I'm not sure why you'd want to support scenarios with "hundreds/thousands of schemes" if there's no real support in the options stack (I'm not sure having hundreds of schemes + hundreds of handlers + hundreds of options instances is really _scalable_).

Not being able to invoke my handler for every path is really annoying for me. As a compromise, would you consider bringing back your per-scheme "promiscuous mode"? (aka `CanHandleRequest` :smile:)

HaoK commented:
I'm not an routing expert, but it looks if we forced a pattern on the callback paths to fit into a route constraint, say `/auth/signin/[scheme]` we could easily plug in something that dispatched dynamically.  There might be other ways as well.

HaoK commented:
@PinpointTownes its going to end up being a question of how easy we want to make it to have a global handler like that.  We certainly could have a handleAllRequests flag on AuthenticationScheme that defaults to false.

HaoK commented:
Started cleaning up some of the legacy handler processing. 

`AuthenticateResult.Nothing/None()` is the replacement for `Skip()` since we no longer really 'skip' anything.  The concept only exists a bit in the Remote/ProcessRequest codepath which keeps the concept of ResponseHandled/StopProcessing (the new name for context.SkipToNextMiddleware() which now returns a `None()/Nothing` result

I'm considering introducing a new ProcessRequestResult which would let us elimate the funky request processing that shows up in AuthenticateResult today since that's the main return type for Auth as a whole.

PinpointTownes commented:
> We certainly could have a handleAllRequests flag on AuthenticationScheme that defaults to false.

That would be much appreciated :sweat_smile: 

> The concept only exists a bit in the Remote/ProcessRequest codepath which keeps the concept of ResponseHandled/StopProcessing (the new name for context.SkipToNextMiddleware() which now returns a None()/Nothing result

Hum, but then, what's the difference between `HandleResponse()` and `StopProcessing()`?

HaoK commented:
Difference between HandledResponse and StopProcessing is whether ProcessRequest returns true/false, same as today's handledResponse/Skip.  Controls whether the middleware continues to invoke other handlers or not via the ghetto routing :)

PinpointTownes commented:
I'm confused now.

When `StopProcessing` is called from an event, `HandleRequestAsync` is supposed to return `false`, right? (e.g https://github.com/aspnet/Security/blob/d80b0d3f26a85bfa2125fe4f9003fcad8cf851b9/src/Microsoft.AspNetCore.Authentication.Impl/RemoteAuthenticationHandler.cs#L116)

Yet, returning `false` asks the authentication middleware to continue invoking the remaining authentication middleware and possibly, invoking the rest of the ASP.NET Core pipeline: https://github.com/aspnet/Security/blob/d80b0d3f26a85bfa2125fe4f9003fcad8cf851b9/src/Microsoft.AspNetCore.Authentication.Impl/AuthenticationMiddleware.cs#L49-L52

So AFAICT, your new `StopProcessing()` method actually means "skip the rest of the authentication handler but continue invoking the other middleware and the rest of the pipeline"? IMHO, `Skip` was way clearer.

HaoK commented:
Fair enough, it was mostly the SkipToNextMiddleware that I was renaming, since the emphasis should just be that its skipping any further processing in this handler.  What happens after that is the normal execution.

We can keep it as Continue/Handled/Skip as the three states which correspond to (false, true, false for the AuthMiddleware invoking the rest of the pipeline)

HaoK commented:
The confusing part is that StopProcessing is the correct name from the perspective of the individual handler's usage of this state, since it lives in the BaseControlContext.  I'll leave that method name still as ProcessingCompleted since for both HandledResponse + Skipped, that means true for the purposes of the handler who is using the context.  All of this smells a bit to me which is why I'm playing around with this area right now...

HaoK commented:
Ok I revereted the StopProcessing state back to Skip.  So the net result of these changes are:

AuthenticateResult.Skip => AuthenticateResult.None to signify no failure, no ticket, nothing
BaseControlContext.CheckEventResult => ProcessingCompleted which is what all of the handlers call on event contexts to determine if they should stop procesing.
Skip remains the verb on the event context objects.

HaoK commented:
Note: Identity can consider breaking the dependency on Http with this change, as it can switch to using IAuthenticationService directly without going through HttpContext.  The only other dependency would be on using the RequestAborted as the default cancellation token.  I'm not sure anyone is actually using that functionality the cancellation tokens today, so we could either remove it, or make it settable in some way.

HaoK commented:
Some implications if we try to use the routing middleware instead of our own ghetto routing:
- How to expose the route configuration to the schemes /signin-{Scheme} by default? Scheme a required paramater and AuthHandler would simply lookup the scheme handler and invoke ProcessRequest
```
    schemeBuilder.AddRoute("/signin-{Scheme}"); // Route that would work for all schemes?
    schemeBuilder.AddRoute("/fixedCustom"); // A fixed route would specify the current scheme
```

- We'd have to find some way to plug into the mvc routes, i.e.
```
    app.UseAuthentication();
    app.UseMvc(routes => {
         routes.MapRoutes("mvcstuff");
         routes.MapAuthenticationRoutes(); // uses the configured auth schemes to add routes
    });
```

Or Without MVC, this overload could add routing and configure it?
```
    app.UseAuthentication(routes => { }); // automatically would call routes.MapAuthenticationRoutes()
```

- We'd lose the ability to 'Skip' and let the normal request processing continue.  Once routing selects a route handler, the middleware chain is broken.  So there would be no support for an auth handler to inspect all requests (have to add your own middleware)

HaoK commented:
See https://github.com/aspnet/Security/pull/1144 for an iteration using IOptionsFactory which provides named options (and potentially validation).  It also split the ProcessRequest into a different interface IAuthenticationRequestHandler which is used instead of CallbackPaths to determine which schemes get called by the authentication middleware

Tratcher commented on src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationScheme.cs:
```diff
@@ -0,0 +1,38 @@
+    {
+        public AuthenticationScheme(string name, Type handlerType, IEnumerable<PathString> callbackPaths, IReadOnlyDictionary<string, object> settings)
+        {
+            Name = name ?? throw new ArgumentNullException(nameof(name));
```
@davidfowl has rejected any use of this new syntax. @Eilon has not made an official ruling...

Eilon replied:
We did make a ruling on this and we hate it. No throw expressions please 😄

HaoK replied:
Blame VS :)  I'll expand these again

Tratcher commented on src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationScheme.cs:
```diff
@@ -0,0 +1,38 @@
+        public Type HandlerType { get; }
+
+        /// <summary>
+        /// If true, the AuthenticationMiddleware will call the handler's HandleRequestAsync method.
```
Wrong comment?

HaoK replied:
Many of the comments were already obsolete, we'll need to do a full scrub as part of this

Tratcher commented on src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationScheme.cs:
```diff
@@ -0,0 +1,38 @@
+
+        // Holds things like the configured options instances for the handler
+        // Also replacement for AuthenticationDescription
+        public IReadOnlyDictionary<string, object> Settings { get; }
```
This is really weird. Are we using it for anything besides Options and Description? Why not just expose those as first class properties?

HaoK replied:
No, but I nuked AuthenticationDescription which was similar and unused mostly.  We can put the DisplayName as a property on the scheme itself or inside of here like Description used to...

Tratcher commented on src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationTicket.cs:
```diff
@@ -0,0 +1,42 @@
+        /// <summary>
+        /// Gets the claims-principal with authenticated user identities.
+        /// </summary>
+        public ClaimsPrincipal Principal{ get; private set; }
```
formatting

Tratcher commented on src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationTicket.cs:
```diff
@@ -0,0 +1,42 @@
+        public AuthenticationTicket(ClaimsPrincipal principal, AuthenticationProperties properties, string authenticationScheme)
+        {
+            AuthenticationScheme = authenticationScheme;
+            Principal = principal ?? throw new ArgumentNullException(nameof(principal));
```
rejected syntax

Tratcher commented on src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationToken.cs:
```diff
@@ -0,0 +1,12 @@
+    public class AuthenticationToken
+    {
+        public string Name { get; set; }
+        public string Value { get; set; }
```
Make immutable and set via constructor?

Tratcher commented on src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationOptions.cs:
```diff
@@ -0,0 +1,69 @@
+
+        public IDictionary<string, AuthenticationSchemeBuilder> SchemeMap { get; } = new Dictionary<string, AuthenticationSchemeBuilder>(); // case sensitive?
+
+        public void AddScheme(string name, Action<AuthenticationSchemeBuilder> configureBuilder)
```
Nevermind, I found it on IAuthenticationSchemeProvider

HaoK replied:
Yeah its a bit weird to remove it here, but we could add it if some one wanted to rip out/replace someone else's scheme during configure services timeframe...

HaoK commented on src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationOptions.cs:
```diff
@@ -0,0 +1,69 @@
+            SchemeMap[name] = builder;
+        }
+
+        public void ConfigureScheme(string name, Action<AuthenticationSchemeBuilder> configureBuilder)
```
Its for identity/other people to configure schemes...  this uglyness is why I went with named options, you should probably look at https://github.com/aspnet/Security/pull/1144 which is effectively iteration 3 and the current branch I'm working on.  I'll address these issues overthere

HaoK commented on src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticationOptions.cs:
```diff
@@ -0,0 +1,69 @@
+            }
+        }
+
+        public IDictionary<string, AuthenticationSchemeBuilder> SchemeMap { get; } = new Dictionary<string, AuthenticationSchemeBuilder>(); // case sensitive?
```
Yeah I'm not sure what it should be, do we want to be case sensitive?  Seems more annoying than helpful

Tratcher replied:
They have always been Ordinal. No reason to change them right now.

HaoK commented on samples/SocialSample/Startup.cs:
```diff
@@ -329,20 +330,20 @@ public void Configure(IApplicationBuilder app, ILoggerFactory loggerfactory)
+                // var user = await context.AuthenticateAsync(MicrosoftAccountDefaults.AuthenticationScheme);
 
                 // Deny anonymous request beyond this point.
                 if (user == null || !user.Identities.Any(identity => identity.IsAuthenticated))
```
its basically equivalent, either services.AuthenticateAsync(context), or context.AuthenticateAsync(), the context methods are just doing GetRequiredService

HaoK commented on src/Microsoft.AspNetCore.Authentication.Abstractions/AuthenticateResult.cs:
```diff
@@ -0,0 +1,90 @@
+            return new AuthenticateResult() { Handled = true };
+        }
+
+        public static AuthenticateResult None()
```
Skip makes sense when its part of the pipeline, like for handle request in the remote case.  It doesn't really make sense when you ask for an identity to get skip...  Empty or Nothing is really what we want here, basically its when there's No cookie, or no token...  NotFound, NoPrincipal also work

HaoK commented on samples/SocialSample/Startup.cs:
```diff
@@ -280,16 +279,18 @@ public void Configure(IApplicationBuilder app, ILoggerFactory loggerfactory)
+                    foreach (var provider in await schemeProvider.GetAllSchemesAsync())
                     {
-                        await context.Response.WriteAsync("<a href=\"?authscheme=" + type.AuthenticationScheme + "\">" + (type.DisplayName ?? "(suppressed)") + "</a><br>");
+                        // REVIEW: we lost access to display name (which is buried in the handler options)
```
Yeah its fallen into a gap, since they hang off the scheme options, but those are configured too late, the schemes need to be registered as part of ConfigureServices.  We can just have the handler set the schemes display name after its options are initialized i guess

Tratcher commented on src/Microsoft.AspNetCore.Authentication.Impl/DefaultAuthenticationService.cs:
```diff
@@ -0,0 +1,139 @@
+            await handler.SignOutAsync(signOutContext);
+        }
+
+        public virtual Task SignInAsync(HttpContext httpContext, string authenticationScheme, ClaimsPrincipal principal)
```
Why isn't this an extension?

HaoK replied:
Yeah unneeded these already exists as extensions on httpContext, I'll remove

Tratcher commented on src/Microsoft.AspNetCore.Authentication.Impl/DefaultAuthenticationService.cs:
```diff
@@ -0,0 +1,139 @@
+            return SignInAsync(httpContext, authenticationScheme, principal, properties: null);
+        }
+
+        public virtual Task ForbidAsync(HttpContext httpContext, string authenticationScheme)
```
Extension?

HaoK replied:
Yup nuked

Tratcher commented on src/Microsoft.AspNetCore.Authentication.Impl/DefaultClaimsTransformation.cs:
```diff
@@ -0,0 +1,19 @@
+    /// <summary>
+    /// Default claims transformation is a no-op.
+    /// </summary>
+    public class DefaultClaimsTransformation : IClaimsTransformation
```
NoOpClaimsTransformation

Tratcher commented on src/Microsoft.AspNetCore.Authentication.Impl/DefaultAuthenticationService.cs:
```diff
@@ -0,0 +1,139 @@
+
+namespace Microsoft.AspNetCore.Authentication
+{
+    public class DefaultAuthenticationService : IAuthenticationService
```
AuthenticationService

HaoK commented on src/Microsoft.AspNetCore.Authentication.Impl/DefaultAuthenticationSchemeProvider.cs:
```diff
@@ -0,0 +1,133 @@
+    /// <summary>
+    /// Builds the actual AuthenticationScheme instances from the AuthenticationOptions.
+    /// </summary>
+    public class DefaultAuthenticationSchemeProvider : IAuthenticationSchemeProvider
```
Good call, dropped Default prefix across the board

HaoK commented:
Replaced with https://github.com/aspnet/Security/pull/1151

HaoK commented on samples/SocialSample/Startup.cs:
```diff
@@ -45,100 +44,68 @@ public Startup(IHostingEnvironment env)
-                SaveTokens = true,
+                o.AppId = Configuration["facebook:appid"];
+                o.AppSecret = Configuration["facebook:appsecret"];
+                o.Scope.Clear();
```
Clears removed

HaoK commented on samples/SocialSample/Startup.cs:
```diff
@@ -329,20 +330,20 @@ public void Configure(IApplicationBuilder app, ILoggerFactory loggerfactory)
-                // var user = await context.Authentication.AuthenticateAsync(AuthenticationManager.AutomaticScheme);
+                // var user = await context.AuthenticateAsync(AuthenticationManager.AutomaticScheme);
 
                 // This is what [Authorize(ActiveAuthenticationSchemes = MicrosoftAccountDefaults.AuthenticationScheme)] calls
```
That was a name that slipped through, was supposed to be renamed to AuthenticationScheme to match the policy :(
